pkg/controller/certificates/trigger/policies/gatherer_test.go

```diff
@@ -0,0 +1,273 @@
+		wantRequest            *cmapi.CertificateRequest
+		wantSecret             *corev1.Secret
+		wantErr                string
+	}{
+		"the returned secret should stay nil when it is not found": {
+			givenCert: gen.Certificate("cert-1",
+				gen.SetCertificateSecretName("secret-1"),
+				gen.SetCertificateUID("uid-1"),
+			),
+			mockSecretLister: listers.FakeSecretListerFrom(listers.NewFakeSecretLister(),
+				listers.SetFakeSecretNamespaceListerGet(nil, apierrors.NewNotFound(cmapi.Resource("Secret"), "secret-1")),
+			),
+			wantRequestsListerCall: func(mock *listers.CertificateRequestListerMock) {},
+			wantSecret:             nil,
+		},
+		"should return an error when getsecret returns an unexpect error that isnt not_found": {
+			givenCert: gen.Certificate("cert-1",
+				gen.SetCertificateSecretName("secret-1"),
+				gen.SetCertificateUID("uid-1"),
+			),
+			mockSecretLister: listers.FakeSecretListerFrom(listers.NewFakeSecretLister(),
+				listers.SetFakeSecretNamespaceListerGet(nil, fmt.Errorf("error that is not a not_found error")),
+			),
+			wantRequestsListerCall: func(mock *listers.CertificateRequestListerMock) {},
+			wantErr:                "error that is not a not_found error",
+		},
+		"the returned certificaterequest should stay nil when the list function returns nothing": {
+			givenCert: gen.Certificate("cert-1",
+				gen.SetCertificateSecretName("secret-1"),
+				gen.SetCertificateUID("uid-1"),
+			),
+			mockSecretLister: listers.FakeSecretListerFrom(listers.NewFakeSecretLister(),
+				listers.SetFakeSecretNamespaceListerGet(nil, nil),
+			),
+			wantRequestsListerCall: func(mock *listers.CertificateRequestListerMock) {},
+			wantRequest:            nil,
+		},
+		"should find the certificaterequest that matches revision and owner": {
+			givenCert: gen.Certificate("cert-1",
+				gen.SetCertificateUID("uid-7"),
+				gen.SetCertificateSecretName("secret-1"),
+				gen.SetCertificateRevision(7),
+			),
+			mockSecretLister: listers.FakeSecretListerFrom(listers.NewFakeSecretLister(),
+				listers.SetFakeSecretNamespaceListerGet(nil, nil),
+			),
+			wantRequestsListerCall: func(mock *listers.CertificateRequestListerMock) {
+				mock.
+					CallCertificateRequests("default-unit-test-ns").
+					CallList("").
+					ReturnList([]*cmapi.CertificateRequest{
+						gen.CertificateRequest("cr-4",
+							gen.AddCertificateRequestOwnerReferences(certRef("cert-1", "uid-4")),
+							gen.AddCertificateRequestAnnotations(map[string]string{
+								"cert-manager.io/certificate-revision": "4",
+							}),
+						),
+						gen.CertificateRequest("cr-7",
+							gen.AddCertificateRequestOwnerReferences(certRef("cert-1", "uid-7")),
+							gen.AddCertificateRequestAnnotations(map[string]string{
+								"cert-manager.io/certificate-revision": "7",
+							}),
+						),
+						gen.CertificateRequest("cr-9",
+							gen.AddCertificateRequestOwnerReferences(certRef("cert-1", "uid-9")),
+						),
+					}, nil)
+			},
```

Comment:
I still don't fully understand the intention of this mock lister.

It doesn't really matter how many times we are calling list, and we only ever make a call with `selector.Everything()`. What we _are_ interested in, is the behaviour of what happens depending on what the lister _returns_, whether that be an error or otherwise. 

I think these mocks are adding quite a bit of complexity here, and isn't consistent with how we are doing tests elsewhere. I'm in strong favour of replacing these mocks with a simple fake client.

I'd be keen to hear other peeps opinions. 

/cc @munnerz @wallrj @jakexks @meyskens

Reply:
> These mocks are adding quite a bit of complexity

Good point

> These mocks aren't consistent with how we are doing tests elsewhere

True also

In my defense, I still make the point that checking that the right input arguments have been given to the mock is important; and in order to check these input arguments, I need to make sure that the mocked call was actually called, otherwise my input data did not have the effect I thought it had.

Reply:
If checking the input is something that is deemed important, I think this can be done on a test by test basis. Something like this is a bit simpler.

You can add this helper function to the file:

```go
func baseLister(nscmlister cmlisters.CertificateRequestNamespaceLister, expNamespace string) func(t *testing.T) cmlisters.CertificateRequestLister {
	return func(t *testing.T) cmlisters.CertificateRequestLister {
		return fakelisters.NewCertificateRequests().
			WithCertificateRequests(func(namespace string) cmlisters.CertificateRequestNamespaceLister {
				if namespace != expNamespace {
					t.Errorf("unexpected namespace, exp=%s, got=%s", expNamespace, namespace)
				}

				return nscmlister
			})
	}
}
```

Your test case definitions can then focus on the return output:

```go
nslister: fakelisters.NewCertificateRequestNamespace().WithList(func (_ selector.Selector) ([]*cmapi.CertificateRequest, error) {
....
}),
```

Then in the Run function:

```go
			g := &Gatherer{
				CertificateRequestLister: baseLister(test.nslister, test.expNamespace)(t),
                                 ....
			}
```

What do you think?

Reply:
Yes, this implementation also works quite well: I implemented it in a new branch [unit-test-dataforcertificate-josh](https://github.com/maelvls/cert-manager/find/unit-test-dataforcertificate-josh).

Take a look at:

- [listers/certificaterequest.go](https://github.com/maelvls/cert-manager/blob/unit-test-dataforcertificate-josh/test/unit/listers/certificaterequest.go)
- [gatherer_test.go](https://github.com/maelvls/cert-manager/blob/unit-test-dataforcertificate-josh/pkg/controller/certificates/trigger/policies/gatherer_test.go). 

I didn't do the "checking that the call was actually made" yet though (otherwise the whole checking input is useless)

Reply:
Can another pair of eyes take a look to the current PR vs. the alternative approach proposed by Josh?

Thanks! @jakexks @meyskens @wallrj